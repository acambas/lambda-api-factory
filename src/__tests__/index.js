import apiFactory from '../index'

test('test simple api service', async () => {
  const service = async () => {
    return 5
  }
  const handler = apiFactory(service)
  const callback = jest.fn()
  await handler({}, {}, callback)
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body: '5',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 200,
      },
    ],
  ])
})

test('test simple api service with body', async () => {
  const service = async () => {
    return 5
  }
  const handler = apiFactory(service)
  const callback = jest.fn()
  await handler(
    {
      queryStringParameters: {},
      pathParameters: {},
      body: JSON.stringify({ test: 'test value' }),
      headers: { 'Content-Type': 'application/json' },
    },
    {},
    callback
  )
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body: '5',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 200,
      },
    ],
  ])
})

test('it should fail because body is in wrong format', async () => {
  const service = async () => {
    return 5
  }
  const handler = apiFactory(service)
  const callback = jest.fn()
  await handler(
    {
      body: '{',
      headers: { 'Content-Type': 'application/json' },
    },
    {},
    callback
  )
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body:
          '{"message":"There was an issue with request body, it cant be parsed into valid JSON"}',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 400,
      },
    ],
  ])
})

test('test simple api service that returns empty result', async () => {
  const service = async () => {
    return
  }
  const handler = apiFactory(service)
  const callback = jest.fn()
  await handler({}, {}, callback)
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body: '{"message":"Resource not found"}',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 404,
      },
    ],
  ])
})

test('test simple api service with sync service', async () => {
  const service = () => {
    return 5
  }
  const handler = apiFactory(service)
  const callback = jest.fn()
  await handler({}, {}, callback)
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body: '5',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 200,
      },
    ],
  ])
})

test('test api service with validation', async () => {
  const service = () => {
    return 5
  }
  const validate = () => {}
  const handler = apiFactory(service, validate)
  const callback = jest.fn()
  await handler({}, {}, callback)
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body: '5',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 200,
      },
    ],
  ])
})

test('test api service with validation that throws exception', async () => {
  const service = () => {
    return 5
  }
  const validate = () => {
    throw new Error('validation issues')
  }
  const handler = apiFactory(service, validate)
  const callback = jest.fn()
  await handler({}, {}, callback)
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body: '{"message":"validation issues"}',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 400,
      },
    ],
  ])
})

test('test api service with validation that throws exception', async () => {
  const service = () => {
    return 5
  }
  const validate = () => {
    throw new Error('validation issues')
  }
  const handler = apiFactory(service, validate)
  const callback = jest.fn()
  await handler({}, {}, callback)
  expect(callback.mock.calls).toEqual([
    [
      null,
      {
        body: '{"message":"validation issues"}',
        headers: { 'Content-Type': 'application/json' },
        statusCode: 400,
      },
    ],
  ])
})

test('test api service with validation that returns error object', async () => {
  const service = () => {
    return 5
  }
  const validate = () => {
    return { msg: 'some value is wrong :)' }
  }
  const handler = apiFactory(service, validate)
  const callback = jest.fn()
  await handler({}, {}, callback)
  expect(callback.mock.calls).toMatchSnapshot()
})
