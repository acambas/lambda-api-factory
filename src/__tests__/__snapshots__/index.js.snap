// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test api service with validation that returns error object 1`] = `
Array [
  Array [
    null,
    Object {
      "body": "{\\"message\\":\\"Validation error: \\\\n              {\\\\n  \\\\\\"msg\\\\\\": \\\\\\"some value is wrong :)\\\\\\"\\\\n}\\"}",
      "headers": Object {
        "Content-Type": "application/json",
      },
      "statusCode": 400,
    },
  ],
]
`;
